### PR TITLE
Add the possibility of reading with more dataitens:

### PR DESCRIPTION
**Pull request by rtorresmatheus**
Add the possibility of reading with more dataitens:

- Added full-time flow to all tags;
- Added new event conditions for change, upper threshold and lower threshold;
- Modified the type of variable responsible for sensor id.

### CODE REVIEW

File: src/device.c [hunk 1]
```diff
@@ -527,14 +531,12 @@ int device_send_config(void)
 	return rc;
 }
 
-void device_publish_data_list(struct l_queue *sensor_id_list)
+void device_publish_data(struct l_queue *sensor_id_list)
```
Comment (luisfelipemisil) [inline src/device.c:534]: l_quere means queue list?

File: src/event.c [hunk 2]
```diff
@@ -123,23 +121,79 @@ static bool is_higher_than_threshold(knot_value_type value,
...
+	timeout_cb();
+	l_timeout_modify(to, *time_sec);
+}
+
+static int comp_change_lower_flag(knot_value_type current_val,
+				  knot_value_type sent_val,
+				  knot_value_type lower_limit,
+				  int value_type)
+{
+	int rc = KNOT_ERR_INVALID;
```
Comment (luisfelipemisil) [inline src/event.c:135]: What rc means?

File: src/event.c [hunk 3]
```diff
@@ -123,23 +121,79 @@ static bool is_higher_than_threshold(knot_value_type value,
...
+
+	if (!is_value_equal(current_val, sent_val, value_type))
+		if (is_lower_than_threshold(current_val,
+				lower_limit, value_type))
+			rc = 0;
```
Comment (luisfelipemisil) [inline src/event.c]: The same here, what does rc means?

File: src/event.c [hunk 4]
```diff
@@ -123,23 +121,79 @@ static bool is_higher_than_threshold(knot_value_type value,
...
+
+	return rc;
```
Comment (luisfelipemisil) [inline src/event.c:142]: The same here, what does rc means?

File: src/event.c [hunk 5]
```diff
@@ -148,43 +202,69 @@ int event_check_value(knot_event event, knot_value_type current_val,
...
-	if (is_change_flag_set(event.event_flags) &&
-			!is_value_equal(current_val, sent_val, value_type))
-		rc = 1;
-	else if (is_lower_flag_set(event.event_flags) &&
-			is_lower_than_threshold(current_val,
+	if (event.event_flags) {
+		switch (event.event_flags) {
+		case KNOT_EVT_FLAG_CHANGE:
+			if (!is_value_equal(current_val, sent_val, value_type))
+				rc = 0;
```
Comment (luisfelipemisil) [inline src/event.c]: I think the number zero should be replaced by a variable with an explicit name.

File: src/event.c [hunk 6]
```diff
@@ -148,43 +202,69 @@ int event_check_value(knot_event event, knot_value_type current_val,
...
+			break;
+		case KNOT_EVT_FLAG_LOWER_THRESHOLD:
+			if (!is_lower_than_threshold(current_val,
 				event.lower_limit, value_type))
-		rc = 1;
-	else if (is_upper_flag_set(event.event_flags) &&
-			is_higher_than_threshold(current_val,
+				rc = 0;
```
Comment (luisfelipemisil) [inline src/event.c]: what does rc means? And I think the number zero should be replaced by a variable with an explicit name

File: src/event.c [hunk 7]
```diff
@@ -148,43 +202,69 @@ int event_check_value(knot_event event, knot_value_type current_val,
...
+		case KNOT_EVENT_FLAG_MAX:
+			if (!comp_all_flag(current_val,
+					event.lower_limit,
+					event.upper_limit,
+					sent_val,
+					value_type))
+				rc = 0;
+			break;
+		default:
+			rc = -1;
```
Comment (luisfelipemisil) [inline src/event.c]: I think the number -1 should be replaced by a variable with an explicit name

File: src/event.c [hunk 8]
```diff
@@ -148,43 +202,69 @@ int event_check_value(knot_event event, knot_value_type current_val,
...
-		struct l_timeout *to = l_timeout_create(event.time_sec,
-							on_sensor_to,
-							data,
-							l_free);
-		l_queue_push_head(sensor_timeouts, to);
-	}
-}
-
-int event_start(timeout_cb_t cb)
+int event_start(timeout_cb_t cb, int time_sec)
```
Comment (luisfelipemisil) [inline src/event.c:258]: what does cb means?

File: src/event.c [hunk 9]
```diff
@@ -195,8 +275,6 @@ int event_start(timeout_cb_t cb)
 void event_stop(void)
 {
 	active = false;
-	if (sensor_timeouts) {
-		l_queue_destroy(sensor_timeouts, timeout_destroy);
-		sensor_timeouts = NULL;
-	}
+	if (timeout_to)
+		l_timeout_remove(timeout_to);
```
Comment (luisfelipemisil) [inline src/event.c:279]: what does to means?

File: src/properties.c [hunk 10]
```diff
@@ -217,20 +218,20 @@ static int set_data_properties(struct knot_thing *thing,
 
 	storage_read_key_int(fd, group_id, DATA_IP_ELEMENT_SIZE,
 			&element_size_aux);
-	if (element_size_aux < 0)
+	if (element_size_aux == -EINVAL)
 		return -EINVAL;
 
 	data_item_aux.element_size = element_size_aux;
 
 	storage_read_key_int(fd, group_id, DATA_REG_ADDRESS,
```
Comment (luisfelipemisil) [inline src/properties.c:226]: what does fd means?

File: src/properties.c [hunk 11]
```diff
@@ -556,16 +550,16 @@ static int set_data_items(struct knot_thing *thing,
 
 error:
 	l_strfreev(data_item_group);
-
 	return -EINVAL;
 }
 
 static int set_driver_properties(struct knot_thing *thing, int fd)
```
Comment (luisfelipemisil) [inline src/properties.c:556]: what does fd means?

File: src/device.c [hunk 12]
```diff
@@ -227,7 +222,6 @@ static void on_driver_connected(void *user_data)
 static int on_driver_poll_receive(int id)
 {
 	struct knot_data_item *data_item_aux;
-	struct l_queue *list;
 	int rc;
```
Comment (luisfelipemisil) [inline src/device.c:225]: what does rc means?

File: src/device.c [hunk 13]
```diff
@@ -236,6 +230,9 @@ static int on_driver_poll_receive(int id)
 		return -EINVAL;
 
 	rc = driver->read(data_item_aux);
+	if (rc < 0)
```
Comment (luisfelipemisil) [inline src/device.c]: what does rc means? And this zero, what does the zero means? I think the number zero should be replaced by a variable with a explicit name

File: src/device.c [hunk 14]
```diff
@@ -236,6 +230,9 @@ static int on_driver_poll_receive(int id)
...
+		return -EPERM;
+
 	if (rawtime == -1) {
```
Comment (luisfelipemisil) [inline src/device.c]: I think the number -1 should be replaced by a variable with a explicit name

File: src/device.c [hunk 15]
```diff
@@ -253,16 +250,22 @@ static int on_driver_poll_receive(int id)
...
-		data_item_aux->sent_val = data_item_aux->current_val;
-		list = l_queue_new();
-		l_queue_push_head(list, &id);
-		sm_input_event(EVT_PUB_DATA, list);
-		l_queue_destroy(list, NULL);
-	}
+			      data_item_aux->schema.value_type) < 0)
+		return -EACCES;
 
 	return rc;
```
Comment (luisfelipemisil) [inline src/device.c:256]: what does rc means?

File: src/device.c [hunk 16]
```diff
@@ -467,15 +474,12 @@ int device_clear_credentials_on_file(void)
 int device_start_event(void)
 {
 	int rc;
```
Comment (luisfelipemisil) [inline src/device.c:476]: what does rc means?

File: src/device.c [hunk 17]
```diff
@@ -467,15 +474,12 @@ int device_clear_credentials_on_file(void)
...
-
-	rc = event_start(on_event_timeout);
+	rc = event_start(on_event_timeout, thing.time_sec);
 	if (rc < 0) {
```
Comment (luisfelipemisil) [inline src/device.c]: what does rc means? And this zero, what does the zero means? I think the number zero should be replaced by a variable with a explicit name

File: src/device.c [hunk 18]
```diff
@@ -467,15 +474,12 @@ int device_clear_credentials_on_file(void)
...
 		l_error("Failed to start event");
 		return rc;
 	}
 
-	l_hashmap_foreach(thing.data_items, foreach_event_add_data_item, NULL);
-
 	return 0;
```
Comment (luisfelipemisil) [inline src/device.c]: What does the zero means? I think the number zero should be replaced by a variable with a explicit name

File: src/device.c [hunk 19]
```diff
@@ -527,14 +531,12 @@ int device_send_config(void)
 	return rc;
```
Comment (luisfelipemisil) [inline src/device.c:531]: what does rc means?

File: src/event.c [hunk 20]
```diff
@@ -123,23 +121,79 @@ static bool is_higher_than_threshold(knot_value_type value,
 
 static void on_sensor_to(struct l_timeout *to, void *data)
 {
-	struct data_item_timeout *data_item_to_info = data;
+	int *time_sec = data;
+
+	timeout_cb();
+	l_timeout_modify(to, *time_sec);
```
Comment (luisfelipemisil) [inline src/event.c:127]: what does to means? And this L on the function l_timeout, same question.

File: src/event.c [hunk 21]
```diff
@@ -123,23 +121,79 @@ static bool is_higher_than_threshold(knot_value_type value,
 
 static void on_sensor_to(struct l_timeout *to, void *data)
 {
-	struct data_item_timeout *data_item_to_info = data;
+	int *time_sec = data;
+
+	timeout_cb();
```
Comment (luisfelipemisil) [inline src/event.c:126]: what does cb means?

File: src/event.c [hunk 22]
```diff
@@ -123,23 +121,79 @@ static bool is_higher_than_threshold(knot_value_type value,
...
+
+	return rc;
+}
+
+static int comp_change_upper_flag(knot_value_type current_val,
+				  knot_value_type sent_val,
+				  knot_value_type upper_limit,
+				  int value_type)
+{
+	int rc = KNOT_ERR_INVALID;
```
Comment (luisfelipemisil) [inline src/event.c:150]: what does rc means?

File: src/event.c [hunk 23]
```diff
@@ -123,23 +121,79 @@ static bool is_higher_than_threshold(knot_value_type value,
...
+
+	if (!is_value_equal(current_val, sent_val, value_type))
+		if (!is_higher_than_threshold(current_val,
+				upper_limit, value_type))
+			rc = 0;
```
Comment (luisfelipemisil) [inline src/event.c]: what does rc means?

File: src/event.c [hunk 24]
```diff
@@ -123,23 +121,79 @@ static bool is_higher_than_threshold(knot_value_type value,
...
+
+	return rc;
```
Comment (luisfelipemisil) [inline src/event.c:157]: what does rc means?

File: src/event.c [hunk 25]
```diff
@@ -123,23 +121,79 @@ static bool is_higher_than_threshold(knot_value_type value,
...
+}
+
+static int comp_upper_lower_flag(knot_value_type current_val,
+			     knot_value_type lower_limit,
+			     knot_value_type upper_limit,
+			     int value_type)
+{
+		int rc = KNOT_ERR_INVALID;
```
Comment (luisfelipemisil) [inline src/event.c:165]: what does rc means?

File: src/event.c [hunk 26]
```diff
@@ -123,23 +121,79 @@ static bool is_higher_than_threshold(knot_value_type value,
...
+
+	if (!is_higher_than_threshold(current_val, upper_limit, value_type))
+		if (!is_lower_than_threshold(current_val, lower_limit,
+					value_type))
+			rc = 0;
```
Comment (luisfelipemisil) [inline src/event.c]: what does rc means?

File: src/event.c [hunk 27]
```diff
@@ -123,23 +121,79 @@ static bool is_higher_than_threshold(knot_value_type value,
...
 
-	timeout_cb(data_item_to_info->id);
-	l_timeout_modify(to, data_item_to_info->timeout_sec);
+	return rc;
```
Comment (luisfelipemisil) [inline src/event.c:172]: what does rc means?

File: src/event.c [hunk 28]
```diff
@@ -123,23 +121,79 @@ static bool is_higher_than_threshold(knot_value_type value,
...
 {
-	struct l_timeout *to = data;
+	int rc = KNOT_ERR_INVALID;
+
+	if (!is_value_equal(current_val, sent_val, value_type))
+		if (!is_higher_than_threshold(current_val, upper_limit,
+					value_type))
+			if (!is_lower_than_threshold(current_val, lower_limit,
+					value_type))
+				rc = 0;
```
Comment (luisfelipemisil) [inline src/event.c]: what does rc means?

File: src/event.c [hunk 29]
```diff
@@ -123,23 +121,79 @@ static bool is_higher_than_threshold(knot_value_type value,
...
 
-	l_timeout_remove(to);
+	return rc;
```
Comment (luisfelipemisil) [inline src/event.c:190]: what does rc means?

File: src/event.c [hunk 30]
```diff
@@ -123,23 +121,79 @@ static bool is_higher_than_threshold(knot_value_type value,
...
 }
 
 int event_check_value(knot_event event, knot_value_type current_val,
 		      knot_value_type sent_val, int value_type)
 {
-	int rc;
+	int rc = KNOT_ERR_INVALID;
```
Comment (luisfelipemisil) [inline src/event.c:196]: what does rc means?

File: src/event.c [hunk 31]
```diff
@@ -148,43 +202,69 @@ int event_check_value(knot_event event, knot_value_type current_val,
...
-	else if (is_upper_flag_set(event.event_flags) &&
-			is_higher_than_threshold(current_val,
+				rc = 0;
+			break;
+		case KNOT_EVT_FLAG_UPPER_THRESHOLD:
+			if (!is_higher_than_threshold(current_val,
 				event.upper_limit, value_type))
-		rc = 1;
-	else
+				rc = 0;
```
Comment (luisfelipemisil) [inline src/event.c]: what does rc means?

File: src/event.c [hunk 32]
```diff
@@ -148,43 +202,69 @@ int event_check_value(knot_event event, knot_value_type current_val,
...
+			break;
+		case KNOT_EVENT_CHANGE_UPPER_FLAG:
+			if (!comp_change_upper_flag(current_val, sent_val,
+						   event.upper_limit,
+						   value_type))
+				rc = 0;
```
Comment (luisfelipemisil) [inline src/event.c]: what does rc means?

File: src/event.c [hunk 33]
```diff
@@ -148,43 +202,69 @@ int event_check_value(knot_event event, knot_value_type current_val,
...
+			break;
+		case KNOT_EVENT_CHANGE_LOWER_FLAG:
+			if (!comp_change_lower_flag(current_val, sent_val,
+				event.lower_limit,
+				value_type))
+				rc = 0;
```
Comment (luisfelipemisil) [inline src/event.c]: what does rc means?

File: src/event.c [hunk 34]
```diff
@@ -148,43 +202,69 @@ int event_check_value(knot_event event, knot_value_type current_val,
...
+			break;
+		case KNOT_EVENT_UPPER_LOWER_FLAG:
+			if (!comp_upper_lower_flag(current_val,
+						event.lower_limit,
+						event.upper_limit,
+						value_type))
+				rc = 0;
```
Comment (luisfelipemisil) [inline src/event.c]: what does rc means?

File: src/event.c [hunk 35]
```diff
@@ -148,43 +202,69 @@ int event_check_value(knot_event event, knot_value_type current_val,
...
+			break;
+		case KNOT_EVENT_FLAG_MAX:
+			if (!comp_all_flag(current_val,
+					event.lower_limit,
+					event.upper_limit,
+					sent_val,
+					value_type))
+				rc = 0;
```
Comment (luisfelipemisil) [inline src/event.c]: what does rc means? And I think the number 0 should be replaced by a variable with an explicit name

File: src/event.c [hunk 36]
```diff
@@ -148,43 +202,69 @@ int event_check_value(knot_event event, knot_value_type current_val,
...
+			break;
+		default:
+			rc = -1;
+			break;
+		}
+	} else
 		rc = 0;
```
Comment (luisfelipemisil) [inline src/event.c]: what does rc means? And I think the number 0 should be replaced by a variable with an explicit name

File: src/event.c [hunk 37]
```diff
@@ -148,43 +202,69 @@ int event_check_value(knot_event event, knot_value_type current_val,
...
 
 	return rc;
```
Comment (luisfelipemisil) [inline src/event.c:255]: what does rc means?

File: src/event.c [hunk 38]
```diff
@@ -148,43 +202,69 @@ int event_check_value(knot_event event, knot_value_type current_val,
...
+	int *time_sec_aux = l_new(int, 1);
+
+	*time_sec_aux = time_sec;
+	timeout_to = l_timeout_create(time_sec,
+				      on_sensor_to,
+				      time_sec_aux,
+				      l_free);
+	if (!timeout_to)
 		return -ENOMSG;
 	timeout_cb = cb;
```
Comment (luisfelipemisil) [inline src/event.c:269]: what does cb means?

File: src/event.h [hunk 39]
```diff
@@ -14,10 +14,9 @@
...
  */
 
-typedef void (*timeout_cb_t)(int);
+typedef void (*timeout_cb_t)();
 
 int event_check_value(knot_event event, knot_value_type current_val,
 		      knot_value_type sent_val, int value_type);
-int event_start(timeout_cb_t cb);
-void event_add_data_item(int id, knot_event event);
+int event_start(timeout_cb_t cb, int time_sec);
```
Comment (luisfelipemisil) [inline src/event.h:21]: what does cb means?

File: src/iface-opc-ua.c [hunk 40]
```diff
@@ -239,6 +233,7 @@ int iface_opc_ua_config(struct knot_data_item *data_item,
 int iface_opc_ua_read_data(struct knot_data_item *data_item)
 {
 	UA_StatusCode rc = UA_STATUSCODE_GOOD;
```
Comment (luisfelipemisil) [inline src/iface-opc-ua.c:235]: what does rc means?

File: src/iface-opc-ua.c [hunk 41]
```diff
@@ -250,16 +245,22 @@ int iface_opc_ua_read_data(struct knot_data_item *data_item)
 				    data_item->value_type_size, &tmp);
 	}
 
-	if (rc != UA_STATUSCODE_GOOD) {
+	if (rc == UA_STATUSCODE_BADCONNECTIONCLOSED) {
 		l_error("Failed to read from OPC UA: %#010x",
 			rc);
```
Comment (luisfelipemisil) [inline src/iface-opc-ua.c:250]: what does rc means?

File: src/iface-opc-ua.c [hunk 42]
```diff
@@ -250,16 +245,22 @@ int iface_opc_ua_read_data(struct knot_data_item *data_item)
...
 		l_io_destroy(opc_ua_io);
 		opc_ua_io = NULL;
-	} else {
+		rental = -EINTR;
+	} else if (rc == UA_STATUSCODE_GOOD) {
```
Comment (luisfelipemisil) [inline src/iface-opc-ua.c:254]: what does rc means?

File: src/iface-opc-ua.c [hunk 43]
```diff
@@ -250,16 +245,22 @@ int iface_opc_ua_read_data(struct knot_data_item *data_item)
...
 		memcpy(&data_item->current_val, &tmp, sizeof(tmp));
+	} else {
+		l_error("Failed to read from OPC UA for Data_%d: %#010x",
+			data_item->sensor_id, rc);
```
Comment (luisfelipemisil) [inline src/iface-opc-ua.c:258]: what does rc means?

File: src/properties.c [hunk 44]
```diff
@@ -354,7 +355,7 @@ static int get_lower_limit(int fd, char *group_id, int value_type,
 static int set_limit(int fd, const char *group_id, const char *key,
 		     int value_type, knot_value_type value)
 {
-	int rc;
+	int rc = KNOT_ERR_INVALID;
```
Comment (luisfelipemisil) [inline src/properties.c:358]: what does cb means?

File: src/properties.c [hunk 45]
```diff
@@ -413,19 +414,12 @@ static int set_event(struct knot_thing *thing, int fd, char *group_id,
...
 	rc = storage_read_key_int(fd, group_id, EVENT_CHANGE, &aux);
 	if (rc > 0)
 		event_aux.event_flags |= KNOT_EVT_FLAG_CHANGE;
 
 	rc = knot_event_is_valid(event_aux.event_flags,
 				 schema.value_type,
-				 event_aux.time_sec,
 				 &event_aux.lower_limit,
 				 &event_aux.upper_limit);
 	if (rc)
```
Comment (luisfelipemisil) [inline src/properties.c:425]: what does cb means?

File: src/properties.c [hunk 46]
```diff
@@ -500,10 +494,10 @@ static int set_sensor_id(struct knot_thing *thing, int fd, char *group_id,
 static int set_data_items(struct knot_thing *thing,
 			  int fd)
 {
-	int rc;
-	int i;
+	int rc = KNOT_ERR_INVALID;
```
Comment (luisfelipemisil) [inline src/properties.c:497]: what does cb means?

File: src/properties.c [hunk 47]
```diff
@@ -582,6 +576,7 @@ static int set_driver_properties(struct knot_thing *thing, int fd)
 		return -EINVAL;
 	}
 	device_set_protocol_type(thing, protocol);
+	l_free(protocol);
 
 	name_type = storage_read_key_string(fd, THING_GROUP,
```
Comment (luisfelipemisil) [inline src/properties.c:581]: what does fd means?

File: src/properties.c [hunk 48]
```diff
@@ -592,6 +587,7 @@ static int set_driver_properties(struct knot_thing *thing, int fd)
 		} else
 			device_set_driver_name_type(thing, name_type);
 	}
+	l_free(name_type);
 
 	login = storage_read_key_string(fd, THING_GROUP,
```
Comment (luisfelipemisil) [inline src/properties.c:592]: what does fd means?

File: src/properties.c [hunk 49]
```diff
@@ -612,6 +609,7 @@ static int set_driver_properties(struct knot_thing *thing, int fd)
 		} else
 			device_set_driver_password(thing, password);
 	}
+	l_free(password);
 
 	path_certificate = storage_read_key_string(fd, THING_GROUP,
```
Comment (luisfelipemisil) [inline src/properties.c:614]: what does fd means?

File: src/properties.c [hunk 50]
```diff
@@ -623,6 +621,7 @@ static int set_driver_properties(struct knot_thing *thing, int fd)
 			device_set_driver_path_certificate(thing,
 							   path_certificate);
 	}
+	l_free(path_certificate);
 
 	security_policy = storage_read_key_string(fd, THING_GROUP,
```
Comment (luisfelipemisil) [inline src/properties.c:626]: what does fd means?

File: src/properties.c [hunk 51]
```diff
@@ -634,6 +633,7 @@ static int set_driver_properties(struct knot_thing *thing, int fd)
 			device_set_driver_security_policy(thing,
 							   security_policy);
 	}
+	l_free(security_policy);
 
 	path_private_key = storage_read_key_string(fd, THING_GROUP,
```
Comment (luisfelipemisil) [inline src/properties.c:638]: what does fd means?

File: src/properties.c [hunk 52]
```diff
@@ -645,37 +645,35 @@ static int set_driver_properties(struct knot_thing *thing, int fd)
...
+	if (time_sec == -EINVAL)
+		return -EINVAL;
+	device_set_driver_time_sec(thing, time_sec);
+
+	storage_read_key_int(fd, THING_GROUP, DRIVER_SECURITY, &security);
+	if (security == -EINVAL)
 		return -EINVAL;
 	device_set_driver_security(thing, security);
 
 	storage_read_key_int(fd, THING_GROUP, DRIVER_ID, &id);
```
Comment (luisfelipemisil) [inline src/properties.c:660]: what does fd means?

File: src/properties.c [hunk 53]
```diff
@@ -645,37 +645,35 @@ static int set_driver_properties(struct knot_thing *thing, int fd)
...
+	l_free(path_private_key);
 
-	storage_read_key_int(fd, THING_GROUP, DRIVER_ID, &security);
-	if (security < 0)
+	storage_read_key_int(fd, THING_GROUP, DRIVER_TIME_SEC, &time_sec);
+	if (time_sec == -EINVAL)
+		return -EINVAL;
+	device_set_driver_time_sec(thing, time_sec);
+
+	storage_read_key_int(fd, THING_GROUP, DRIVER_SECURITY, &security);
```
Comment (luisfelipemisil) [inline src/properties.c:655]: what does fd means?

File: src/properties.c [hunk 54]
```diff
@@ -645,37 +645,35 @@ static int set_driver_properties(struct knot_thing *thing, int fd)
 			device_set_driver_path_private_key(thing,
 							   path_private_key);
 	}
+	l_free(path_private_key);
 
-	storage_read_key_int(fd, THING_GROUP, DRIVER_ID, &security);
-	if (security < 0)
+	storage_read_key_int(fd, THING_GROUP, DRIVER_TIME_SEC, &time_sec);
```
Comment (luisfelipemisil) [inline src/properties.c:650]: what does fd means?

File: src/properties.c [hunk 55]
```diff
@@ -645,37 +645,35 @@ static int set_driver_properties(struct knot_thing *thing, int fd)
...
 
-	l_free(name_type);
-	l_free(login);
-	l_free(password);
-	l_free(protocol);
-	l_free(security_policy);
-	l_free(path_private_key);
-	l_free(path_certificate);
-
 	return 0;
```
Comment (luisfelipemisil) [inline src/properties.c]: I think the number zero should be replaced by a variable with a explicit name.

File: src/properties.c [hunk 56]
```diff
@@ -804,17 +802,6 @@ static int update_event_data_item(int fd, const char *group_id,
 {
 	int rc;
```
Comment (luisfelipemisil) [inline src/properties.c:803]: Whats does the rc means?

File: src/properties.c [hunk 57]
```diff
@@ -804,17 +802,6 @@ static int update_event_data_item(int fd, const char *group_id,
...
-		if (rc < 0) {
-			l_error("Failed to set new time sec");
-			return rc;
-		}
-	} else if (!storage_has_unit(fd, group_id, EVENT_TIME_SEC)) {
-		storage_remove_key(fd, group_id, EVENT_TIME_SEC);
-	}
-
 	if (event->event_flags & KNOT_EVT_FLAG_CHANGE) {
 		rc = storage_write_key_int(fd, group_id, EVENT_CHANGE,
```
Comment (luisfelipemisil) [inline src/properties.c:806]: what does rc means?

File: src/storage.c [hunk 58]
```diff
@@ -109,7 +109,10 @@ static int check_valid_groups(char **all_groups, int index)
 	regex_t preg;
 	int i;
```
Comment (luisfelipemisil) [inline src/storage.c]: what does i means?

File: tests/mocks/fake-device.c [hunk 59]
```diff
@@ -75,12 +75,7 @@ int device_clear_credentials_on_file(void)
 	return 0;
```
Comment (luisfelipemisil) [inline tests/mocks/fake-device.c:75]: I think the number zero should be replaced by a variable with an explicit name

**Review by luisfelipemisil — COMMENTED** [review]
Really nice code, very clean as always, with only a few numbers and variables I don't understand, but maybe could be something in the protocol that I dont know.

**Review by AntonioJanael — APPROVED** [review]
Great job!

File: src/device.c (hunk 1 above)
Comment (rtorresmatheus) [reply src/device.c:534]: Not. All variables and functions that have the prefix l_ come from the ell library.

File: src/event.c (hunk 2 above)
Comment (rtorresmatheus) [reply src/event.c:135]: Return calculation. It is a common way of abbreviating.

File: src/event.c (hunk 8 above)
Comment (rtorresmatheus) [reply src/event.c:258]: Callback Function. It is a common way of abbreviating.

File: src/event.c (hunk 9 above)
Comment (rtorresmatheus) [reply src/event.c:279]: Function in lib ell to remove timeout.

File: src/properties.c (hunk 10 above)
Comment (rtorresmatheus) [reply src/properties.c:226]: File destination. It is a common way of abbreviating.

Comment (rtorresmatheus) [reply src/properties.c:226]: File destination. It is a common way of abbreviating.

**Review by AntonioJanael — APPROVED** [review]

**Pull request merged**